### PR TITLE
Keras-like API for training and evaluation

### PR DESCRIPTION
**Pull request by hkvision**
## What changes were proposed in this pull request?

compile, fit, evaluate and predict for Keras-like API models on Scala and Python with doc in each method.
Refine LeNet example on Scala and Python using compile and fit for training and can achieve the same result.

## How was this patch tested?

Jenkins
Unit tests for compile fit evaluate predict on both Scala and Python.

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

**Comment by hkvision** [comment]
@jason-dai have a review?

File: pyspark/bigdl/nn/keras/layer.py [hunk 1]
```diff
@@ -64,7 +66,36 @@ def jvm_class_constructor(self):
         return name
 
 
-class Sequential(TSequential, InferShape):
+class Training(JavaValue):
+    def __init__(self, bigdl_type="float"):
+        self.bigdl_type = bigdl_type
+
+    def compile(self, optimizer, loss, metrics=None):
+        init_engine()
```
Comment (hkvision) [inline pyspark/bigdl/nn/keras/layer.py]: Wondering where shall we put `init_engine`, `redirect_spark_logs` and `show_bigdl_info_logs` as in the new API these should be transparent from the user.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDLKeras.scala [hunk 2]
```diff
@@ -664,4 +666,54 @@ class PythonBigDLKeras[T: ClassTag](implicit ev: TensorNumeric[T]) extends Pytho
...
+  def compile(
+    module: Model[T],
+    optimizer: String,
+    loss: String,
+    metrics: JList[String]): Unit = {
+    module.compile(optimizer, loss, metrics.asScala.toArray)
+  }
+
+  def compile(
+    module: Model[T],
```
Comment (hkvision) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDLKeras.scala]: How to do the wrapping for both Sequential and Model in one def?

**Comment by hkvision** [comment]
jenkins passed http://172.168.2.101:8080/job/BigDL-PR-Validation-v3/257/

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Topology.scala [hunk 3]
```diff
@@ -28,7 +29,11 @@ import scala.reflect.ClassTag
 
 class Model[T: ClassTag](private val _inputs : Seq[ModuleNode[T]],
       private val _outputs : Seq[ModuleNode[T]])(implicit ev: TensorNumeric[T])
-  extends StaticGraph[T](_inputs, _outputs, None, false) {
+  extends StaticGraph[T](_inputs, _outputs, None, false) with Training[T] {
+
+  override var model: Module[T] = this
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Topology.scala]: It's weird to add the `model` here.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/KerasUtils.scala [hunk 4]
```diff
@@ -106,4 +108,46 @@ object KerasUtils {
     }
   }
 
+  private[keras] def toBigDLCriterion[T : ClassTag](loss: String)
+    (implicit ev: TensorNumeric[T]): Criterion[T] = {
+    loss.toLowerCase() match {
+      case "categorical_crossentropy" => CategoricalCrossEntropy[T]()
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/KerasUtils.scala]: we should move the unittest from python to scala as well with the mapping logic

File: pyspark/bigdl/optim/keras/training.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+#
+
+from bigdl.util.common import *
+
+
+class Training(JavaValue):
+    def __init__(self, bigdl_type="float"):
+        self.bigdl_type = bigdl_type
+
+    def compile(self, optimizer, loss, metrics=None):
```
Comment (zhichao-li) [inline pyspark/bigdl/optim/keras/training.py]: Should they keep the same hierarchy same as Scala? i.e put them into "KerasLayer"

Comment (zhichao-li) [reply pyspark/bigdl/optim/keras/training.py]: Maybe "KerasLayer" also not a good place for this as ideally we only allow "Model" and "Sequential" to have compile and fit.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Topology.scala (hunk 3 above)
Comment (hkvision) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Topology.scala]: This is to override the model for compile and fit. Any suggestion for modification?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/KerasUtils.scala (hunk 4 above)
Comment (hkvision) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/KerasUtils.scala]: Newly added criterions already have unit tests. I will add the old ones in Scala.

File: pyspark/bigdl/nn/keras/topology.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+class KerasModel(KerasLayer):
+    def compile(self, optimizer, loss, metrics=None):
+        callBigDlFunc(self.bigdl_type, "compile",
+                      self.value,
+                      optimizer,
+                      loss,
+                      metrics)
+
+    def fit(self, x, y=None, batch_size=32, nb_epoch=10, validation_data=None, distributed=True):
+        init_engine()
```
Comment (zhichao-li) [inline pyspark/bigdl/nn/keras/topology.py]: remove this

File: pyspark/bigdl/nn/keras/topology.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        redire_spark_logs()
+        show_bigdl_info_logs()
+        if distributed:
+            if isinstance(x, np.ndarray) and isinstance(y, np.ndarray):
+                training_data = to_sample_rdd(x, y)
+                if validation_data:
+                    validation_data = to_sample_rdd(*validation_data)
+            elif isinstance(x, RDD) and not y:
```
Comment (zhichao-li) [inline pyspark/bigdl/nn/keras/topology.py]: add "DataSet" here as well

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Topology.scala [hunk 8]
```diff
@@ -16,27 +16,164 @@
...
+  }
+
+  /**
+   * Train a model for a fixed number of epochs on a dataset.
+   * @param x Training data.
+   * @param batchSize Number of samples per gradient update.
+   * @param nbEpoch Number of iterations to train.
+   * @param validationData Null if validation is not configured.
+   */
+  def fit(x: RDD[Sample[T]], batchSize: Int = 32, nbEpoch: Int = 10,
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Topology.scala]: I think we should only keep two `fit` methods here: 1) accept DataSet 2) accept RDD[Sample] which is just a syntax sugar of `DataSet`

File: pyspark/bigdl/nn/keras/topology.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+                          batch_size,
+                          nb_epoch,
+                          val_x,
+                          val_y,
+                          multiprocessing.cpu_count())
+
+    def evaluate(self, x, y=None, batch_size=32):
+        if isinstance(x, np.ndarray) and isinstance(y, np.ndarray):
+            evaluation_data = to_sample_rdd(x, y)
+        elif isinstance(x, RDD) and not y:
```
Comment (zhichao-li) [inline pyspark/bigdl/nn/keras/topology.py]: same as fit?

File: pyspark/bigdl/nn/keras/topology.py [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+        return callBigDlFunc(self.bigdl_type, "evaluate",
+                             self.value,
+                             evaluation_data,
+                             batch_size)
+
+    def predict(self, x, distributed=True):
+        if is_distributed:
+            if isinstance(x, np.ndarray):
+                features = to_sample_rdd(x, np.zeros([x.shape[0]]))
+            elif isinstance(x, RDD):
```
Comment (zhichao-li) [inline pyspark/bigdl/nn/keras/topology.py]: same as fit?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Topology.scala [hunk 11]
```diff
@@ -16,27 +16,144 @@
...
+               batchSize: Int)
+      (implicit ev: TensorNumeric[T]): Array[(ValidationResult, ValidationMethod[T])] = {
+    require(this.vMethods != null, "Evaluation metrics haven't been set yet")
+    this.evaluate(x, this.vMethods, Some(batchSize))
+  }
+
+  /**
+   * Evaluate a model in LOCAL mode.
+   */
+  def evaluate(x: LocalDataSet[MiniBatch[T]])
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Topology.scala]: let's refactor evaluate and predict the same as `fit`

**Comment by hkvision** [comment]
jenkins passed http://172.168.2.101:8080/view/PR-V(new)/job/BigDL-PR-Validation-v3/377/

**Pull request merged**